gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTaskLauncher.java

```diff
@@ -0,0 +1,121 @@
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.gobblin.util.GobblinProcessBuilder;
+import org.apache.gobblin.util.SystemPropertiesWrapper;
+
+import static org.apache.gobblin.cluster.SingleTaskRunnerMainOptions.CLUSTER_CONFIG_FILE_PATH;
+import static org.apache.gobblin.cluster.SingleTaskRunnerMainOptions.JOB_ID;
+import static org.apache.gobblin.cluster.SingleTaskRunnerMainOptions.WORK_UNIT_FILE_PATH;
+
+
+class SingleTaskLauncher {
+  private static final Logger logger = LoggerFactory.getLogger(SingleTaskLauncher.class);
+
+  private final GobblinProcessBuilder processBuilder;
+  private final SystemPropertiesWrapper propertiesWrapper;
+  private final Path clusterConfigFilePath;
+
+  SingleTaskLauncher(final GobblinProcessBuilder processBuilder,
+      final SystemPropertiesWrapper propertiesWrapper, final Path clusterConfigFilePath) {
+    this.processBuilder = processBuilder;
+    this.propertiesWrapper = propertiesWrapper;
+    this.clusterConfigFilePath = clusterConfigFilePath;
+  }
+
+  Process launch(final String jobId, final Path workUnitFilePath)
+      throws IOException {
+    final SingleTaskLauncher.CmdBuilder cmdBuilder = this.new CmdBuilder(jobId, workUnitFilePath);
+    final List<String> command = cmdBuilder.build();
+    logger.info("Launching a task process. cmd: " + command);
+    final Process taskProcess = this.processBuilder.build(command);
+
+    return taskProcess;
+  }
+
+  private class CmdBuilder {
+    private final String jobId;
+    private final Path workUnitFilePath;
+    private final List<String> cmd = new ArrayList<>();
+
+    private CmdBuilder(final String jobId, final Path workUnitFilePath) {
+      this.jobId = jobId;
+      this.workUnitFilePath = workUnitFilePath;
+    }
+
+    List<String> build() {
+      addJavaBin();
+      addClassPath();
+      addClassName();
+      addOptions();
+      return this.cmd;
+    }
+
+    private void addClassName() {
+      final String runnerClassName = SingleTaskRunnerMain.class.getCanonicalName();
+      this.cmd.add(runnerClassName);
+    }
+
+    private void addJavaBin() {
+      final String javaHomeDir = SingleTaskLauncher.this.propertiesWrapper.getJavaHome();
+      final Path javaBinPath = Paths.get(javaHomeDir, "bin", "java");
+      this.cmd.add(javaBinPath.toString());
+    }
+
+    private void addClassPath() {
+      this.cmd.add("-cp");
+      final String classPath = SingleTaskLauncher.this.propertiesWrapper.getJavaClassPath();
+      this.cmd.add(classPath);
+    }
+
+    private void addOptions() {
+      addClusterConfigPath();
+      addJobId();
+      addWorkUnitPath();
+    }
+
+    private void addClusterConfigPath() {
+      this.cmd.add(formatParam(CLUSTER_CONFIG_FILE_PATH));
+      this.cmd.add(SingleTaskLauncher.this.clusterConfigFilePath.toString());
```

Comment:
in all addX() methods, shouldn't we check for availability of param values they are adding? else we might have empty or null strings being attached to param keys

Reply:
The assumption is that these values are valid when they get here. It expects the validation to be done at the place where these values are created.

What additional values would validation here bring?